fvm/fvm_test.go

```diff
@@ -4371,3 +4371,101 @@ func TestScriptAccountKeyMutationsFailure(t *testing.T) {
+				privateKeys, err := testutil.GenerateAccountPrivateKeys(3)
+				require.NoError(t, err)
+
+				// Bootstrap a ledger, creating accounts with the provided private keys and the root account.
+				accounts, err := testutil.CreateAccounts(vm, view, ps, privateKeys, chain)
+				require.NoError(t, err)
+				account1 := accounts[0]
+				account2 := accounts[1]
+				account3 := accounts[2]
+
+				contractA := "pub contract A {}"
+				contractB := fmt.Sprintf(`
+					import A from 0x%s 
+					pub contract B {}
+				`, account1.String())
+				contractC := fmt.Sprintf(`
+					import A from 0x%s 
+					import B from 0x%s 
+					pub contract C {}
+				`, account1.String(), account2.String())
+
+				txBodyA := testutil.CreateContractDeploymentTransaction("A", contractA, account1, chain)
+				txBodyB := testutil.CreateContractDeploymentTransaction("B", contractB, account2, chain)
+				txBodyC := testutil.CreateContractDeploymentTransaction("C", contractC, account3, chain)
+				txBodyA.SetProposalKey(chain.ServiceAddress(), 0, 0)
+				txBodyA.SetPayer(chain.ServiceAddress())
+				txBodyB.SetProposalKey(chain.ServiceAddress(), 0, 1)
+				txBodyB.SetPayer(chain.ServiceAddress())
+				txBodyC.SetProposalKey(chain.ServiceAddress(), 0, 2)
+				txBodyC.SetPayer(chain.ServiceAddress())
+
+				err = testutil.SignPayload(txBodyA, account1, privateKeys[0])
+				require.NoError(t, err)
+
+				err = testutil.SignEnvelope(txBodyA, chain.ServiceAddress(), unittest.ServiceAccountPrivateKey)
+				require.NoError(t, err)
+
+				tx := fvm.Transaction(txBodyA, 0)
+				err = vm.Run(ctx, tx, view, programs.NewEmptyPrograms())
+				require.NoError(t, err)
+				require.NoError(t, tx.Err)
+
+				err = testutil.SignPayload(txBodyB, account2, privateKeys[1])
+				require.NoError(t, err)
+
+				err = testutil.SignEnvelope(txBodyB, chain.ServiceAddress(), unittest.ServiceAccountPrivateKey)
+				require.NoError(t, err)
+
+				tx = fvm.Transaction(txBodyB, 0)
+				err = vm.Run(ctx, tx, view, programs.NewEmptyPrograms())
+				require.NoError(t, err)
+				require.NoError(t, tx.Err)
+
+				err = testutil.SignPayload(txBodyC, account3, privateKeys[2])
+				require.NoError(t, err)
+
+				err = testutil.SignEnvelope(txBodyC, chain.ServiceAddress(), unittest.ServiceAccountPrivateKey)
+				require.NoError(t, err)
+
+				tx = fvm.Transaction(txBodyC, 0)
+				err = vm.Run(ctx, tx, view, programs.NewEmptyPrograms())
+				require.NoError(t, err)
+				require.NoError(t, tx.Err)
+
+				script := fvm.Script([]byte(fmt.Sprintf(`
+					import A from 0x%s
+					import B from 0x%s
+					import C from 0x%s
+					pub fun main() {}
+				`, account1.String(), account2.String(), account3.String())))
+
+				scriptCtx := fvm.NewContextFromParent(ctx)
+
+				err = vm.Run(scriptCtx, script, view, ps)
+				require.NoError(t, err)
+				require.NoError(t, script.Err)
+				// we should meter computation for 3 GetPrograms, one for each contract
+				assert.Equal(t, int(script.GasUsed), 900)
```

Comment:
@janezpodhostnik before adding the cache skipping, this is 1400 instead of 900. I am not sure however why it would be 900 here or what the value actually should be in the ideal situation.

Reply:
After some investigating; the interpreter itself calls get program in addition to the checker to enforce that imports are properly loaded. However these should still be skipped by electing not to merge the state, unless we are adding cost to the state of B for the load of A, and then when we load B from C, for example, it counts the cost of both B and A when loading C?

Reply:
Ok after some more thought about this over the weekend I think just skipping merge after the first load of each program is not sufficient if we want to only report 300 units for this test. Here is a slightly simpler example:
```
pub contract A {}
```
```
import A from 0x1
pub contract B {}
```
```
import A from 0x1
import B from 0x2
pub fun main() {}
```
Executing the script here with the changes in this PR does the following computation metering and meter merges:
* GetProgram called for check of `import A` in script, increasing the meter of the script to 100
* SetProgram for A is called, state of checking `A` is merged into state of checking script, increasing no metering because no calls to GetProgram were made while checking A. 
* GetProgram called for check of `import B` in script, increasing the metering of the script to 200
* GetProgram called for check of `import A` in `B`, increasing the metering of B to 100
* SetProgram for B is called, state of checking `B` is merged into state of checking script, increasing metering of the script to 300
* GetProgram called for interpret of `import A` in script, increasing the metering of the script to 400
* GetProgram called for interpret of `import B` in script , increasing the metering of the script to 500

Because the merging occurs during the calls to SetProgram, the import of `A` that gets metered during the check of `B` is added to its meter, which is then merged into the meter of the script, even though we have already merged `A` into the script already.